postgres_exporter: panic: runtime error: integer divide by zero

Config:

```yaml
integrations:
  prometheus_remote_write:
    - url: https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push
      basic_auth:
        username: XXX
        password: YYY
  postgres_exporter:
    enabled: true
    data_source_names:
      - "postgresql://USER:PASSWORD@HOST:5432/DATABASE?sslmode=disable"
  scrape_integrations: true
```

debug agent logs:

```
NOTE: enabling Prometheus subsystem as Integrations are enabled
level=info ts=2021-04-15T17:47:37.978152789Z caller=server.go:229 http=[::]:80 grpc=[::]:9095 msg="server listening on addresses"
time="2021-04-15T17:47:37Z" level=info msg="Established new database connection to \"HOST:5432\"." source="exporter.go:846"
level=debug ts=2021-04-15T17:47:37.984496704Z caller=instance.go:268 agent=prometheus instance=integration/postgres_exporter msg="initializing instance" name=integration/postgres_exporter
level=info ts=2021-04-15T17:47:37.995789202Z caller=wal.go:173 agent=prometheus instance=integration/postgres_exporter msg="replaying WAL, this may take a while" dir=/etc/agent/data/integration/postgres_exporter/wal
level=info ts=2021-04-15T17:47:37.997751619Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=0 maxSegment=20
level=info ts=2021-04-15T17:47:37.99859013Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=1 maxSegment=20
level=info ts=2021-04-15T17:47:37.999584116Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=2 maxSegment=20
level=info ts=2021-04-15T17:47:37.999897664Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=3 maxSegment=20
level=info ts=2021-04-15T17:47:38.000194441Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=4 maxSegment=20
level=info ts=2021-04-15T17:47:38.000457781Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=5 maxSegment=20
level=info ts=2021-04-15T17:47:38.000731408Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=6 maxSegment=20
level=info ts=2021-04-15T17:47:38.001029134Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=7 maxSegment=20
level=info ts=2021-04-15T17:47:38.001329121Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=8 maxSegment=20
level=info ts=2021-04-15T17:47:38.001639976Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=9 maxSegment=20
level=info ts=2021-04-15T17:47:38.001992427Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=10 maxSegment=20
level=info ts=2021-04-15T17:47:38.002269165Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=11 maxSegment=20
level=info ts=2021-04-15T17:47:38.002550853Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=12 maxSegment=20
level=info ts=2021-04-15T17:47:38.003437968Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=13 maxSegment=20
level=info ts=2021-04-15T17:47:38.004839051Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=14 maxSegment=20
level=info ts=2021-04-15T17:47:38.005331228Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=15 maxSegment=20
level=info ts=2021-04-15T17:47:38.005890508Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=16 maxSegment=20
level=info ts=2021-04-15T17:47:38.007379713Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=17 maxSegment=20
level=info ts=2021-04-15T17:47:38.00767526Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=18 maxSegment=20
level=info ts=2021-04-15T17:47:38.007961665Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=19 maxSegment=20
level=info ts=2021-04-15T17:47:38.008239912Z caller=wal.go:220 agent=prometheus instance=integration/postgres_exporter msg="WAL segment loaded" segment=20 maxSegment=20
level=debug ts=2021-04-15T17:47:38.008425903Z caller=manager.go:195 agent=prometheus instance=integration/postgres_exporter component="discovery manager" msg="Starting provider" provider=static/0 subs=[integrations/postgres_exporter]
ts=2021-04-15T17:47:38.012450347Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=info remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Starting WAL watcher" queue=integration/postgres_exporter-a345e6
ts=2021-04-15T17:47:38.012677414Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=info remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Starting scraped metadata watcher"
level=debug ts=2021-04-15T17:47:38.013753336Z caller=instance.go:349 agent=prometheus instance=integration/postgres_exporter msg="running instance" name=integration/postgres_exporter
level=debug ts=2021-04-15T17:47:38.014993708Z caller=manager.go:213 agent=prometheus instance=integration/postgres_exporter component="discovery manager" msg="Discoverer channel closed" provider=static/0
ts=2021-04-15T17:47:38.015430696Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=info remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Replaying WAL" queue=integration/postgres_exporter-a345e6
ts=2021-04-15T17:47:38.015809106Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Tailing WAL" lastCheckpoint= checkpointIndex=0 currentSegment=0 lastSegment=20
ts=2021-04-15T17:47:38.015990188Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=0
ts=2021-04-15T17:47:38.027193765Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=1
time="2021-04-15T17:47:38Z" level=info msg="Semantic Version Changed on \"HOST:5432\": 0.0.0 -> 13.1.0" source="exporter.go:1428"
ts=2021-04-15T17:47:38.037519876Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=2
ts=2021-04-15T17:47:38.049428791Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=3
ts=2021-04-15T17:47:38.059863041Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=4
ts=2021-04-15T17:47:38.070406929Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=5
ts=2021-04-15T17:47:38.080896441Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=6
ts=2021-04-15T17:47:38.091493105Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=7
ts=2021-04-15T17:47:38.101916753Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=8
ts=2021-04-15T17:47:38.112336453Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=9
ts=2021-04-15T17:47:38.122773683Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=10
ts=2021-04-15T17:47:38.133198357Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=11
ts=2021-04-15T17:47:38.143571096Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=12
ts=2021-04-15T17:47:38.154015682Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=13
ts=2021-04-15T17:47:38.164530319Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=14
ts=2021-04-15T17:47:38.175065652Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=15
ts=2021-04-15T17:47:38.185519562Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=16
ts=2021-04-15T17:47:38.195972535Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=17
ts=2021-04-15T17:47:38.206481666Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=18
ts=2021-04-15T17:47:38.217023996Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=19
ts=2021-04-15T17:47:38.227428629Z caller=dedupe.go:112 agent=prometheus instance=integration/postgres_exporter component=remote level=debug remote_name=integration/postgres_exporter-a345e6 url=https://prometheus-blocks-prod-us-central1.grafana.net/api/prom/push msg="Processing segment" currentSegment=20
panic: runtime error: integer divide by zero

goroutine 380 [running]:
github.com/prometheus/prometheus/scrape.(*Target).offset(0xc00029b720, 0x0, 0x4817b9e5513c8189, 0x0)
	/src/agent/vendor/github.com/prometheus/prometheus/scrape/target.go:161 +0xf3
github.com/prometheus/prometheus/scrape.(*scrapeLoop).run(0xc00092fb80, 0x0, 0x0, 0x0)
	/src/agent/vendor/github.com/prometheus/prometheus/scrape/scrape.go:1001 +0x71
created by github.com/prometheus/prometheus/scrape.(*scrapePool).sync
	/src/agent/vendor/github.com/prometheus/prometheus/scrape/scrape.go:511 +0x9ce
```

PostgreSQL version: 

```
PostgreSQL 13.1 on x86_64-pc-linux-musl, compiled by gcc (Alpine 9.3.0) 9.3.0, 64-bit
(1 row)
```

Commenting out ___ helps:

```
#  postgres_exporter:
#    enabled: true
#    data_source_names:
#      - "postgresql://USER:PASSWORD@HOST:5432/DATABASE?sslmode=disable"
```

logs after disabling `postgres_exporter`:

```
NOTE: enabling Prometheus subsystem as Integrations are enabled
level=info ts=2021-04-15T17:43:51.566153367Z caller=server.go:229 http=[::]:80 grpc=[::]:9095 msg="server listening on addresses"
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,13 +3,17 @@ cross-compilation issue, but will return in v0.13.0.
 
 # Main (unreleased)
 
+- [BUGFIX] Ensure defaults are applied to undefined sections in config file.
+  This fixes a problem where integrations didn't work if `prometheus:` wasn't
+  configured. (@rfratto)
+
 # 0.14.0-rc.3 (2021-04-15)
 
 - [ENHANCEMENT] Add  `headers` field in `remote_write` config for Tempo. `headers`
   specifies HTTP headers to forward to the remote endpoint. (@alexbiehl)
 
 - [BUGFIX] Grafana Agent running as a Windows service should start automatically on startup
-  (@mattdurham) 
+  (@mattdurham)
 
 - [BUGFIX] Validate that incoming scraped metrics do not have an empty label
   set or a label set with duplicate labels, mirroring the behavior of

--- a/pkg/config/config.go
+++ b/pkg/config/config.go
@@ -18,6 +18,13 @@ import (
 	"gopkg.in/yaml.v2"
 )
 
+// DefaultConfig holds default settings for all the subsystems.
+var DefaultConfig = Config{
+	// All subsystems with a DefaultConfig should be listed here.
+	Prometheus:   prom.DefaultConfig,
+	Integrations: integrations.DefaultManagerConfig,
+}
+
 // Config contains underlying configurations for the agent
 type Config struct {
 	Server       server.Config              `yaml:"server,omitempty"`
@@ -31,6 +38,13 @@ type Config struct {
 	// to restart.
 	ReloadAddress string `yaml:"-"`
 	ReloadPort    int    `yaml:"-"`
+}
+
+// UnmarshalYAML implements yaml.Unmarshaler.
+func (c *Config) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultConfig
+	type config Config
+	return unmarshal((*config)(c))
 }
 
 // ApplyDefaults sets default values in the config
@@ -103,8 +117,9 @@ func Load(fs *flag.FlagSet, args []string) (*Config, error) {
 // doesn't require having a literal file on disk.
 func load(fs *flag.FlagSet, args []string, loader func(string, bool, *Config) error) (*Config, error) {
 	var (
+		cfg = DefaultConfig
+
 		printVersion    bool
-		cfg             Config
 		file            string
 		configExpandEnv bool
 	)

--- a/pkg/config/config_test.go
+++ b/pkg/config/config_test.go
@@ -6,6 +6,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/grafana/agent/pkg/integrations"
+	"github.com/grafana/agent/pkg/prom"
 	"github.com/grafana/agent/pkg/prom/instance"
 	"github.com/prometheus/common/model"
 	promCfg "github.com/prometheus/prometheus/config"
@@ -119,4 +121,13 @@ prometheus:
 		err := LoadBytes([]byte(cfg), false, &c)
 		require.Error(t, err)
 	})
+}
+
+func TestConfig_Defaults(t *testing.T) {
+	var c Config
+	err := LoadBytes([]byte(`{}`), false, &c)
+	require.NoError(t, err)
+
+	require.Equal(t, prom.DefaultConfig, c.Prometheus)
+	require.Equal(t, integrations.DefaultManagerConfig, c.Integrations)
 }

